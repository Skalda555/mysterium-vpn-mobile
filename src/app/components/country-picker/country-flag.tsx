/*
 * Copyright (C) 2018 The 'MysteriumNetwork/mysterium-vpn-mobile' Authors.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { Icon } from 'native-base'
import * as React from 'react'
import { Image, StyleSheet } from 'react-native'
import colors from '../../styles/colors'
import countries from './countries'

type FlagProps = {
  countryCode: string | null
}

const getCountryImageUri = (code: string): string => {
  if (countries[code]) {
    return countries[code].image
  }

  return ''
}

const CountryFlag: React.SFC<FlagProps> = ({ countryCode }) => {
  if (!countryCode || !countries[countryCode]) {
    return (
      <Icon style={styles.globeIcon} name={'ios-globe'}/>
    )
  }

  return (
    <Image
      source={{ uri: getCountryImageUri(countryCode) }}
      style={styles.countryFlagImage}
    />
  )
}

const styles = StyleSheet.create({
  countryFlagImage: {
    width: 26,
    height: 26,
    borderRadius: 13,
    borderWidth: 0.5,
    borderColor: colors.border
  },
  globeIcon: {
    color: colors.primary
  }
})

export default CountryFlag
