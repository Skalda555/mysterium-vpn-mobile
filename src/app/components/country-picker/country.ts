import { ProposalDTO } from 'mysterium-tequilapi'
import { Proposal } from '../../../libraries/favorite-proposal'

type Country = {
  id: string
  name: string,
  countryCode: string
}

const proposalToCountry = (proposalDto: ProposalDTO): Country => {
  const proposal = new Proposal(proposalDto, false)

  return {
    id: proposalDto.providerId,
    name: proposal.name,
    countryCode: proposal.getCountryCode(proposalDto)
  }
}

const proposalsToCountries = (proposals?: ProposalDTO[]): Country[] => {
  if (!proposals) {
    return []
  }

  return proposals.map(proposalToCountry)
}

export {
  proposalsToCountries,
  Country
}
