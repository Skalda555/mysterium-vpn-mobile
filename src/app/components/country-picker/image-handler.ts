import countries from './countries.json'

const getCountryImageFile = (code: string) => {
  if (countries[code]) {
    return countries[code].image
  }

  return ''
}

export {
  getCountryImageFile
}
