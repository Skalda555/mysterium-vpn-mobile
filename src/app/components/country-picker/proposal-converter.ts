import { ProposalDTO } from 'mysterium-tequilapi'
import { CountryListItem } from './country-picker'
import { Proposal } from '../../../libraries/favorite-proposal'

const getCountryListItemsFromProposals = (proposals?: ProposalDTO[]) => {
  if (!proposals) {
    return []
  }

  const countryList = proposals.map((proposalDto: ProposalDTO): CountryListItem => {
    const proposal = new Proposal(proposalDto, false)

    return {
      id: proposalDto.providerId,
      name: proposal.name,
      countryCode: proposal.getCountryCode(proposalDto)
    }
  })

  return countryList
}

export {
  getCountryListItemsFromProposals
}
