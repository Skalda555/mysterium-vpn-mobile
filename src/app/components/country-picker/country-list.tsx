import {
  Body,
  Button,
  Container,
  Content,
  Header,
  Icon,
  Input,
  Item,
  Left,
  List,
  ListItem,
  Right,
  Text
} from 'native-base'
import React, { ReactNode } from 'react'
import { Platform, StyleSheet } from 'react-native'
import translations from '../../translations'
import { Country } from './country'
import CountryFlag from './country-flag'

type ListProps = {
  items: Country[],
  onClose: () => void,
  onSelect: (country: Country) => void
}

type ListState = {
  countries: Country[]
}

class CountryList extends React.Component<ListProps, ListState> {
  constructor (props: ListProps) {
    super(props)

    this.state = { countries: this.props.items }
  }

  public render (): ReactNode {
    return (
      <Container>
        <Header>
          <Item style={styles.headerItem}>
            <Icon name="ios-search" style={styles.searchIcon}/>
            <Input
              placeholderTextColor={platformStyles.search.inputColor}
              placeholder={translations.COUNTRY_SEARCH}
              onChange={(event) => this.onSearchValueChange(event.nativeEvent.text)}
              style={styles.searchInput}
            />
          </Item>
          <Right>
            <Button transparent={true} onPress={() => this.props.onClose()}>
              <Text>Close</Text>
            </Button>
          </Right>
        </Header>
        <Content>
          <List>
            {this.state.countries.map((country: Country) => this.renderListItem(country))}
          </List>
        </Content>
      </Container>
    )
  }

  private renderListItem (country: Country): ReactNode {
    return (
      <ListItem icon={true} key={country.id} onPress={() => this.props.onSelect(country)}>
        <Left style={styles.flagImage}>
          <CountryFlag countryCode={country.countryCode}/>
        </Left>
        <Body>
        <Text>{country.name}</Text>
        </Body>
      </ListItem>
    )
  }

  private onSearchValueChange (text: string) {
    let countries = this.props.items

    if (!text.trim().length) {
      this.setState({ countries })

      return
    }

    countries = this.props.items.filter((country: Country) => {
      return country.name.toLowerCase().includes(text.toLowerCase())
    })

    this.setState({ countries })
  }
}

const iosSearchColor = '#222'
const androidSearchColor = '#fff'

const platformStyles = {
  search: {
    iconColor: Platform.OS === 'ios' ? iosSearchColor : androidSearchColor,
    inputColor: Platform.OS === 'ios' ? iosSearchColor : androidSearchColor
  }
}

const styles: any = StyleSheet.create({
  headerItem: {
    borderWidth: 0,
    width: '70%',
    borderColor: 'transparent'
  },
  flagImage: {
    width: 26,
    height: 26,
    margin: 0,
    padding: 0
  },
  searchIcon: {
    color: platformStyles.search.iconColor
  },
  searchInput: {
    color: platformStyles.search.inputColor
  }
})

export default CountryList
