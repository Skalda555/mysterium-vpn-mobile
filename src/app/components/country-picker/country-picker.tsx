/*
 * Copyright (C) 2018 The 'MysteriumNetwork/mysterium-vpn-mobile' Authors.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import {
  Col,
  Grid,
  Icon,
  Text
} from 'native-base'
import * as React from 'react'

import { StyleSheet, TouchableOpacity, View } from 'react-native'
import colors from '../../../app/styles/colors'
import { Country } from './country'
import CountryFlag from './country-flag'
import CountryList from './country-list'
import CountryModal from './country-modal'

type PickerProps = {
  items: Country[],
  onSelect: (country: Country) => void,
  placeholder: string
}

type PickerState = {
  modalIsOpen: boolean,
  selectedCountry?: Country
}

class CountryPicker extends React.Component<PickerProps, PickerState> {
  constructor (props: PickerProps) {
    super(props)

    this.state = { modalIsOpen: false }
  }

  public render () {
    return (
      <View style={styles.container}>
        <CountryModal
          isOpen={this.state.modalIsOpen}
          onClose={() => this.closeCountryModal()}
        >
          <CountryList
            items={this.props.items}
            onClose={() => this.closeCountryModal()}
            onSelect={(country: Country) => this.onCountrySelect(country)}
          />
        </CountryModal>

        <View style={styles.countryPicker}>
          <Grid>
            <Col size={85}>
              <TouchableOpacity onPress={() => this.openCountryModal()}>
                <Grid>
                  <Col size={15} style={styles.countryFlagBox}>
                    <CountryFlag countryCode={this.countryCode}/>
                  </Col>

                  <Col size={90} style={styles.countryNameBox}>
                    <Text>{this.countryName}</Text>
                  </Col>

                  <Col size={10} style={styles.arrowBox}>
                    <Icon style={styles.arrowIcon} name="arrow-down"/>
                  </Col>
                </Grid>
              </TouchableOpacity>
            </Col>
            <Col size={15} style={styles.favoritesBox}>
              <Icon style={styles.favoritesIcon} name="md-star-outline"/>
            </Col>
          </Grid>
        </View>
      </View>
    )
  }

  private get countryCode () {
    if (!this.state.selectedCountry) {
      return
    }

    return this.state.selectedCountry.countryCode.toLowerCase()
  }

  private get countryName () {
    if (!this.state.selectedCountry) {
      return this.props.placeholder
    }

    return this.state.selectedCountry.name
  }

  private openCountryModal () {
    this.setState({ modalIsOpen: true })
  }

  private closeCountryModal () {
    this.setState({ modalIsOpen: false })
  }

  private onCountrySelect (country: Country) {
    this.props.onSelect(country)

    this.setState({
      selectedCountry: country,
      modalIsOpen: false
    })
  }
}

const styles = StyleSheet.create({
  container: {
    width: '100%',
    backgroundColor: '#fff'
  },
  countryPicker: {
    justifyContent: 'center',
    alignItems: 'center',
    borderRadius: 4,
    borderWidth: 0.5,
    borderColor: colors.border
  },
  countryFlagBox: {
    justifyContent: 'center',
    alignItems: 'center',
    borderLeftWidth: 0.5,
    borderColor: colors.border,
    borderRightWidth: 0.5
  },
  countryNameBox: {
    justifyContent: 'center',
    alignItems: 'center',
    height: 40
  },
  arrowBox: {
    justifyContent: 'center',
    alignItems: 'center',
    marginRight: 5,
    width: '100%'
  },
  arrowIcon: {
    fontSize: 20,
    marginTop: 4,
    color: colors.primary
  },
  favoritesBox: {
    justifyContent: 'center',
    alignItems: 'center',
    borderLeftWidth: 0.5,
    borderColor: colors.border
  },
  favoritesIcon: {
    color: colors.primary
  }
})

export default CountryPicker
