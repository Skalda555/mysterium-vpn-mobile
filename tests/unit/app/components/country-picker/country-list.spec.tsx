import { render, shallow } from 'enzyme'
import React from 'react'
import { Country } from '../../../../../src/app/components/country-picker/country'
import CountryList from '../../../../../src/app/components/country-picker/country-list'
import consoleOverride from '../../../helpers/console-override'

describe('CountryList', () => {
  consoleOverride()

  let items: Country[] = []
  const onSelect = () => {
    //
  }
  const onClose = () => {
    //
  }

  beforeEach(() => {
    items = []
  })

  it('renders correctly', () => {
    const wrapper = shallow(
      <CountryList
        items={items}
        onSelect={onSelect}
        onClose={onClose}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })

  const renderWrap = () => {
    return render(
      <CountryList
        items={items}
        onSelect={onSelect}
        onClose={onClose}
      />
    )
  }
  it('renders items', () => {
    items = [
      {
        id: '0x1',
        name: 'Lithuania',
        countryCode: 'lt'
      },
      {
        id: '0x2',
        name: 'United States',
        countryCode: 'us'
      }
    ]

    const wrapper = renderWrap()

    expect(wrapper.text()).toContain('Lithuania')
    expect(wrapper.text()).toContain('United States')
  })

  it('has close button', () => {
    const wrapper = renderWrap()

    expect(wrapper.text()).toContain('Close')
  })
})
