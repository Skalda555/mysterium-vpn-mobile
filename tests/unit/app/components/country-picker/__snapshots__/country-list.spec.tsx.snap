// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountryList renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CountryList
    items={Array []}
    onClose={[Function]}
    onSelect={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Styled(Header)>
          <Styled(Item)
            style={
              Object {
                "borderColor": "transparent",
                "borderWidth": 0,
                "width": "70%",
              }
            }
          >
            <Styled(Icon)
              name="ios-search"
              style={
                Object {
                  "color": "#222",
                }
              }
            />
            <Styled(Input)
              onChange={[Function]}
              placeholder="Search countries"
              placeholderTextColor="#222"
              style={
                Object {
                  "color": "#222",
                }
              }
            />
          </Styled(Item)>
          <Styled(Right)>
            <Styled(Button)
              onPress={[Function]}
              transparent={true}
            >
              <Styled(Text)>
                Close
              </Styled(Text)>
            </Styled(Button)>
          </Styled(Right)>
        </Styled(Header)>,
        <Styled(Content)>
          <Styled(List) />
        </Styled(Content)>,
      ],
      "virtual": undefined,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Styled(Item)
              style={
                Object {
                  "borderColor": "transparent",
                  "borderWidth": 0,
                  "width": "70%",
                }
              }
            >
              <Styled(Icon)
                name="ios-search"
                style={
                  Object {
                    "color": "#222",
                  }
                }
              />
              <Styled(Input)
                onChange={[Function]}
                placeholder="Search countries"
                placeholderTextColor="#222"
                style={
                  Object {
                    "color": "#222",
                  }
                }
              />
            </Styled(Item)>,
            <Styled(Right)>
              <Styled(Button)
                onPress={[Function]}
                transparent={true}
              >
                <Styled(Text)>
                  Close
                </Styled(Text)>
              </Styled(Button)>
            </Styled(Right)>,
          ],
          "virtual": undefined,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [
                <Styled(Icon)
                  name="ios-search"
                  style={
                    Object {
                      "color": "#222",
                    }
                  }
                />,
                <Styled(Input)
                  onChange={[Function]}
                  placeholder="Search countries"
                  placeholderTextColor="#222"
                  style={
                    Object {
                      "color": "#222",
                    }
                  }
                />,
              ],
              "style": Object {
                "borderColor": "transparent",
                "borderWidth": 0,
                "width": "70%",
              },
              "virtual": undefined,
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "name": "ios-search",
                  "style": Object {
                    "color": "#222",
                  },
                  "virtual": undefined,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "onChange": [Function],
                  "placeholder": "Search countries",
                  "placeholderTextColor": "#222",
                  "style": Object {
                    "color": "#222",
                  },
                  "virtual": undefined,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
            ],
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <Styled(Button)
                onPress={[Function]}
                transparent={true}
              >
                <Styled(Text)>
                  Close
                </Styled(Text)>
              </Styled(Button)>,
              "virtual": undefined,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <Styled(Text)>
                  Close
                </Styled(Text)>,
                "onPress": [Function],
                "transparent": true,
                "virtual": undefined,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": "Close",
                  "virtual": undefined,
                },
                "ref": null,
                "rendered": "Close",
                "type": [Function],
              },
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <Styled(List) />,
          "virtual": undefined,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [],
            "virtual": undefined,
          },
          "ref": null,
          "rendered": Array [],
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Styled(Header)>
            <Styled(Item)
              style={
                Object {
                  "borderColor": "transparent",
                  "borderWidth": 0,
                  "width": "70%",
                }
              }
            >
              <Styled(Icon)
                name="ios-search"
                style={
                  Object {
                    "color": "#222",
                  }
                }
              />
              <Styled(Input)
                onChange={[Function]}
                placeholder="Search countries"
                placeholderTextColor="#222"
                style={
                  Object {
                    "color": "#222",
                  }
                }
              />
            </Styled(Item)>
            <Styled(Right)>
              <Styled(Button)
                onPress={[Function]}
                transparent={true}
              >
                <Styled(Text)>
                  Close
                </Styled(Text)>
              </Styled(Button)>
            </Styled(Right)>
          </Styled(Header)>,
          <Styled(Content)>
            <Styled(List) />
          </Styled(Content)>,
        ],
        "virtual": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Styled(Item)
                style={
                  Object {
                    "borderColor": "transparent",
                    "borderWidth": 0,
                    "width": "70%",
                  }
                }
              >
                <Styled(Icon)
                  name="ios-search"
                  style={
                    Object {
                      "color": "#222",
                    }
                  }
                />
                <Styled(Input)
                  onChange={[Function]}
                  placeholder="Search countries"
                  placeholderTextColor="#222"
                  style={
                    Object {
                      "color": "#222",
                    }
                  }
                />
              </Styled(Item)>,
              <Styled(Right)>
                <Styled(Button)
                  onPress={[Function]}
                  transparent={true}
                >
                  <Styled(Text)>
                    Close
                  </Styled(Text)>
                </Styled(Button)>
              </Styled(Right)>,
            ],
            "virtual": undefined,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": Array [
                  <Styled(Icon)
                    name="ios-search"
                    style={
                      Object {
                        "color": "#222",
                      }
                    }
                  />,
                  <Styled(Input)
                    onChange={[Function]}
                    placeholder="Search countries"
                    placeholderTextColor="#222"
                    style={
                      Object {
                        "color": "#222",
                      }
                    }
                  />,
                ],
                "style": Object {
                  "borderColor": "transparent",
                  "borderWidth": 0,
                  "width": "70%",
                },
                "virtual": undefined,
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "name": "ios-search",
                    "style": Object {
                      "color": "#222",
                    },
                    "virtual": undefined,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "onChange": [Function],
                    "placeholder": "Search countries",
                    "placeholderTextColor": "#222",
                    "style": Object {
                      "color": "#222",
                    },
                    "virtual": undefined,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
              ],
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <Styled(Button)
                  onPress={[Function]}
                  transparent={true}
                >
                  <Styled(Text)>
                    Close
                  </Styled(Text)>
                </Styled(Button)>,
                "virtual": undefined,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": <Styled(Text)>
                    Close
                  </Styled(Text)>,
                  "onPress": [Function],
                  "transparent": true,
                  "virtual": undefined,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "class",
                  "props": Object {
                    "children": "Close",
                    "virtual": undefined,
                  },
                  "ref": null,
                  "rendered": "Close",
                  "type": [Function],
                },
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <Styled(List) />,
            "virtual": undefined,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": Array [],
              "virtual": undefined,
            },
            "ref": null,
            "rendered": Array [],
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
